lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -33,12 +33,12 @@ def initialize(options={})
       end
 
       def purchase(money, payment, options={})
-        options[:idempotency_key] = nil
         if options[:execute_threed] || options[:threed_dynamic]
           authorize(money, payment, options)
         else
           MultiResponse.run do |r|
             r.process { authorize(money, payment, options) }
+            options[:idempotency_key] = "#{options[:idempotency_key]}-cap" if options[:idempotency_key]
```

Comment:
@jknipp @bayprogrammer The purchase action mutates `options[:idempotency_key]` after `authorize` so that the `capture` request has a unique key. This is why I re-set the key before the second purchase request in the remote test.

I chose to modify it in the exact same way each time, so that repeat purchase requests will lead to a non-unique key for the `capture`. Then the entire purchase action can be repeated without risk of duplication.

Reply:
Ah yes, I saw that. Very elegant solution to the problem!

I'm wondering if it might not be valuable to extend your solution a bit to make the second call with a copy of the input options hash to avoid mutating it in place? Maybe something like:

```ruby
def purchase(money, payment, options={})
  # ...
      r.process { capture(money, r.authorization, capture_options(options)) }
  # ...
end

private

def capture_options(options)
  return options.merge(idempotency_key: "#{options[:idempotency_key]}-cap") if options[:idempotency_key]
  options
end
```

That should have the same effect, but would allow the caller to not have to worry about whether or not the `idempotency_key` they pass in will be mutated between subsequent calls (should they be using the same hash instance for subsequent calls, as we bumped into with the remote test).

Reply:
I like this a lot! 🆒